### PR TITLE
#14527 Add raw maven input for Edit Maven Artifact Dialog

### PR DESCRIPTION
**Pull request by Destrolaric**

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/DriverEditDialog.java [hunk 1]
```diff
@@ -573,7 +573,10 @@ public void widgetSelected(SelectionEvent e) {
     private void addMavenArtifact() {
         EditMavenArtifactDialog fd = new EditMavenArtifactDialog(getShell(), driver, null);
         if (fd.open() == IDialogConstants.OK_ID) {
-            driver.addDriverLibrary(fd.getLibrary(), true);
+            List<DriverLibraryMavenArtifact> artifacts = fd.getArtifacts();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/DriverEditDialog.java]: Redundant variable - can be inlined.

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java [hunk 2]
```diff
@@ -19,107 +19,201 @@
...
+
+    private void parseArtifactText() {
+        try {
+            artifacts.clear();
+            if (REGEX_FOR_GRADLE.matcher(fieldText.getText()).find()) {
+                artifacts.addAll(parseGradle());
+            } else {
+                artifacts.addAll(parseMaven());
+            }
+            setStatus(false, "Artifacts found: " + artifacts.size());
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java]: Let's translate this status message too.

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java [hunk 3]
```diff
@@ -19,107 +19,201 @@
...
+            Button ignoreDependenciesCheckbox = UIUtils.createCheckbox(settingsGroup, "Ignore transient dependencies", "Do not include library dependencies", originalArtifact != null && originalArtifact.isIgnoreDependencies(), 2);
+            ignoreDependenciesCheckbox.addSelectionListener(new SelectionAdapter() {
+                @Override
+                public void widgetSelected(SelectionEvent e) {
+                    ignoreDependencies = ignoreDependenciesCheckbox.getSelection();
+                }
+            });
+
+            Button loadOptionalDependenciesCheckbox = UIUtils.createCheckbox(settingsGroup, "Load optional dependencies", "Include all optional dependencies", originalArtifact != null && originalArtifact.isIgnoreDependencies(), 2);
+            ignoreDependenciesCheckbox.addSelectionListener(new SelectionAdapter() {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java]: A typo. Should be `loadOptionalDependenciesCheckbox` instead of `ignoreDependenciesCheckbox`.

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java [hunk 4]
```diff
@@ -19,107 +19,201 @@
...
+            tabFolder.setLayoutData(gridData);
+            if (originalArtifact == null) {
+                createRawTab(tabFolder);
+            }
+            createManualTab(tabFolder);
+            tabFolder.addSelectionListener(new SelectionAdapter() {
+                @Override
+                public void widgetSelected(SelectionEvent e) {
+                    if (tabFolder.getSelectionIndex() == 1) {
+                        artifacts.clear();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java]: Why is this only relevant to the first tab?

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java [hunk 5]
```diff
@@ -19,107 +19,201 @@
...
+
     @Override
-    protected Control createDialogArea(Composite parent)
-    {
+    protected Control createDialogArea(Composite parent) {
         getShell().setText(UIConnectionMessages.dialog_edit_driver_edit_maven_title);
-
         Composite composite = (Composite) super.createDialogArea(parent);
-        ((GridLayout)composite.getLayout()).numColumns = 2;
+        ((GridLayout) composite.getLayout()).numColumns = 1;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java]: No point in this statement. Parent layout already consists of a single column.

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java [hunk 6]
```diff
@@ -19,107 +19,201 @@
...
+    private boolean loadOptionalDependencies;
+    private final DriverDescriptor driver;
     private Text groupText;
     private Text artifactText;
     private Text classifierText;
     private Combo versionText;
-    private boolean ignoreDependencies;
-    private boolean loadOptionalDependencies;
+    private Text fieldText;
+    private boolean manualSelected;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java]: Please reorder these fields to be more pleasant to look at. Finals on top, then other variables, SWT components afterwards.

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java [hunk 7]
```diff
@@ -19,107 +19,201 @@
...
+import java.io.IOException;
+import java.io.StringReader;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class EditMavenArtifactDialog extends Dialog {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java]: It would be cool to use our `BaseDialog` instead. See for its usages in other places.

File: plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/driver/DriverLibraryMavenArtifact.java [hunk 8]
```diff
@@ -281,7 +281,7 @@ public String getId() {
 
     @Override
     public String getVersion() {
-        if (preferredVersion != null) {
+        if (preferredVersion != null && !preferredVersion.isEmpty()) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/driver/DriverLibraryMavenArtifact.java:284]: I think this condition causes an empty version to be shown in various situations.

**Review by ShadelessFox — CHANGES_REQUESTED** [review]
👍🏻

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/internal/UIConnectionMessages.properties [hunk 9]
```diff
@@ -223,6 +223,9 @@ pref_page_drivers_maven_label_bad_url_tip = Bad Maven repository URL
 pref_page_drivers_maven_button_remove = Remove
 #pref_page_drivers_maven_button_disable = Disable
 pref_page_drivers_maven_button_up = Up
+dialog_edit_driver_edit_maven_raw = Raw
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/internal/UIConnectionMessages.properties:227]: I believe we should come out with a better name.

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java [hunk 10]
```diff
@@ -128,18 +222,152 @@ private void updateButtons() {
...
+                name = matcher.group("shortName");
+                version = matcher.group("shortVersion");
+            }
+
+            if (CommonUtils.isNotEmpty(group) && CommonUtils.isNotEmpty(name) && CommonUtils.isNotEmpty(version)) {
+                DriverLibraryMavenArtifact lib = new DriverLibraryMavenArtifact(EditMavenArtifactDialog.this.driver, DBPDriverLibrary.FileType.jar, "", "");
+                lib.setReference(new MavenArtifactReference(CommonUtils.notEmpty(group), CommonUtils.notEmpty(name), null, CommonUtils.notEmpty(version)));
+                artifacts.add(lib);
+            } else {
+                throw new XMLException("Wrong parameters!");
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java]: I'm pretty confident that we don't parse XML here. Moreover, I believe that the text of the exception should more clearly indicate what's wrong.

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java [hunk 11]
```diff
@@ -128,18 +222,152 @@ private void updateButtons() {
...
+        final Matcher matcher = REGEX_FOR_GRADLE.matcher(fieldText.getText());
+        while (matcher.find()) {
+            final String classifier = matcher.group("group");
+
+            // Try to match long form
+            String group = matcher.group("group");
+            String name = matcher.group("name");
+            String version = matcher.group("version");
+
+            if (!CommonUtils.isNotEmpty(group) && !CommonUtils.isNotEmpty(name) && !CommonUtils.isNotEmpty(version)) {
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java]: 1. Please get rid of these double negations.
2. [It's totally OK for the version to be empty](https://docs.gradle.org/current/userguide/single_versions.html#sec:declaring_without_version). In this case, we should use the latest version (RELEASE version in our definitions)

**Review by flambchap — CHANGES_REQUESTED** [review]
Generally looks nice 👍

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java [hunk 12]
```diff
@@ -19,103 +19,195 @@
...
+            tabFolder.setLayoutData(gd);
+            if (originalArtifact == null) {
+                createRawTab(tabFolder);
+            }
+            createManualTab(tabFolder);
+        }
+        {
+            Group settingsGroup = UIUtils.createControlGroup(composite, UIConnectionMessages.dialog_edit_driver_edit_maven_settings, 1, GridData.FILL_HORIZONTAL, 0);
+
+            Button ignoreDependenciesCheckbox = UIUtils.createCheckbox(settingsGroup, "Ignore transient dependencies", "Do not include library dependencies", originalArtifact != null && originalArtifact.isIgnoreDependencies(), 2);
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java]: Can be localized.

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java [hunk 13]
```diff
@@ -19,103 +19,195 @@
...
+            ignoreDependenciesCheckbox.addSelectionListener(new SelectionAdapter() {
+                @Override
+                public void widgetSelected(SelectionEvent e) {
+                    ignoreDependencies = ignoreDependenciesCheckbox.getSelection();
+                }
+            });
+
+            Button loadOptionalDependenciesCheckbox = UIUtils.createCheckbox(settingsGroup, "Load optional dependencies", "Include all optional dependencies", originalArtifact != null && originalArtifact.isIgnoreDependencies(), 2);
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java]: Can be localized.

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/internal/UIConnectionMessages.properties (hunk 9 above)
Comment (flambchap) [reply plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/internal/UIConnectionMessages.properties:227]: How about "Dependency declaration"?

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java [hunk 14]
```diff
@@ -19,103 +19,204 @@
...
+        try {
+            artifacts.clear();
+            if (REGEX_FOR_GRADLE.matcher(fieldText.getText()).find()) {
+                artifacts.addAll(parseGradle());
+            } else {
+                artifacts.addAll(parseMaven());
+            }
+            setStatus(false, NLS.bind(UIConnectionMessages.dialog_edit_driver_edit_maven_artifacts_count, artifacts.size()));
+        } catch (Exception e) {
+            setStatus(true, e.getMessage());
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java]: This way the status won't be localized.

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java [hunk 15]
```diff
@@ -19,103 +19,204 @@
...
+            log.trace("Error parsing artifact!", e);
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/driver/EditMavenArtifactDialog.java]: This effectively eats the exception since `trace` is disabled by default. I think we need to include this exception in debug log.
